src/serialize/Schema.ts

```diff
@@ -0,0 +1,355 @@
+// The definition of the core 'Schema' type, as well as a set of data-oriented
+// combinators for building up complex schemas.
+
+/**
+ * A 'Schema<T, S>' encapsulates the behavior to:
+ * - Encode values from domain type 'T' into representation type 'S'
+ * - Decode values from representation type 'S' into domain type 'S'
+ * - Validate that a given 'any' value is of type 'S'
+ *
+ * For example, if we have a class 'Person' with data fields 'name: string' and
+ * 'age: number', we cannot encode/decode directly to/from JSON, since instances
+ * of the class also contain methods that can't be serialized. We only want the
+ * type of the JSON representation to contain the data, so we would define a
+ * 'Schema<Person, { name: string; age: number; }' to handle this for us.
+ */
+export interface Schema<T, S> {
+    encode(src: T): S;
+    decode(data: S): T;
+    validate(data: any): data is S;
+}
+
+/**
+ * Extract the representation type from the type of a schema. This is useful in
+ * making schemas that are derived from other nontrivial schemas. For example:
+ *
+ * <pre><code>
+ * const person: Schema<Person, { name: string, age: number } =
+ *     recordOf({ name: aString, age: aNumber });
+ * const people: Schema<Person[], ReprOf<typeof personSchema>[]> =
+ *     arrayOf(personSchema);
+ * </code></pre>
+ */
+export type ReprOf<T> = T extends Schema<T, infer S> ? S : never;
+
+// Function to use when the compiler needs a path to be present, even when it's
+// technically unreachable
+function impossible(): never {
+    throw Error("This code should be unreachable");
+}
+
+// The identity function, for trivial schemas
+const id = <T>(x: T): T => x;
+
+type NonEmptyArray<T> = [T, ...T[]];
```

Comment:
`type NonEmptyArray<T> = [T, ...T[]];` is now defined in two different places in our code (the other is exported from `src/util/Arrays.ts`). If you tell me that it's better to repeat it here than to import it I trust you, but it seemed like it was worth mentioning.

Reply:
This is intentional; this module is supposed to exist independent of everything else, without imports.